pkg/kubelet/cm/topologymanager/policy_single_numa_node.go

```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+						filtered = append(filtered, hint)
+					}
+				}
+			}
+			filteredResourcesHints = append(filteredResourcesHints, filtered)
+		}
+	}
+	return filteredResourcesHints
+}
+
 func (p *singleNumaNodePolicy) mergeProvidersHints(providersHints []map[string][]TopologyHint) TopologyHint {
 	// Set the default hint to return from this function as an any-NUMANode
 	// affinity with an unpreferred allocation. This will only be returned if
 	// no better hint can be found when merging hints from each hint provider.
 	defaultAffinity, _ := bitmask.NewBitMask(p.numaNodes...)
 	defaultHint := TopologyHint{defaultAffinity, false}
+
+	// Loop through all provider hints and save an accumulated list of the
+	// hints returned by each hint provider. If no hints are provided, assume
+	// that provider has no preference for topology-aware allocation.
+	var allResourcesHints [][]TopologyHint
+	for _, hints := range providersHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with any resource, " +
+				"skipping.")
+			continue
+		}
+
+		// Otherwise, accumulate the hints for each resource type into allProviderHints.
+		for resource := range hints {
+			if hints[resource] == nil {
+				klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with resource "+
+					"'%s', skipping.", resource)
+				continue
+			}
+
+			if len(hints[resource]) == 0 {
+				klog.Infof("[topologymanager] Hint Provider has no possible NUMA affinities for resource '%s'",
+					resource)
+				// return defaultHint which will fail pod admission
+				return defaultHint
+			}
+			klog.Infof("[topologymanager] TopologyHints for resource '%v': %v", resource, hints[resource])
+			allResourcesHints = append(allResourcesHints, hints[resource])
+		}
+	}
+	// In case allProviderHints length is zero it means that we have no
+	// preference for NUMA affinity. In that case update default hint preferred
+	// to true to allow scheduling.
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No preference for NUMA affinity from all providers")
+		defaultHint.Preferred = true
+		return defaultHint
+	}
+
+	allResourcesHints = p.filterHints(allResourcesHints)
+	// if no hints or there is a resource with empty hints after filtering, then policy cannot be satisfied
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No hints that align to a single NUMA node.")
+		return defaultHint
+	}
+	for _, hints := range allResourcesHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] No hints that align to a single NUMA node for resource.")
+			return defaultHint
+		}
+	}
+
+	found, match := p.getHintMatch(allResourcesHints)
+	if found {
+		klog.Infof("[topologymanager] single-numa-node policy match: %v", match)
+		return match
+	}
+	klog.Infof("[topologymanager] single-numa-node no match: %v", defaultHint)
 	return defaultHint
 }
```

Comment:
I'm not a fan of how this branches and returns in two places. Since we have a helper function, I would expect `getHintMatch()` to just return whatever our result without the need to inspect it.  If no good match is found, it should return `TopologyHint{}`, which is what we want to return anyway in the case of "no match".